src/ast.rs

```diff
@@ -3891,11 +3298,11 @@
     if top_type == Type::job_list {
         let mut list = pops.allocate::<JobList>();
         pops.populate_list(&mut *list, true /* exhaust_stream */);
-        ast.top = list;
+        ast.top = Box::new(NodeEnum::from(*list));
```

Comment:
I believe the only place where we really need the generic `From` impl is in the `as_node()` and `as_mut_node()`.

If possible we should not use it in cases like this where we know the concrete type,
because that gives  more information. So I'd write it like

```
ast.top = Box::new(NodeEnum::Branch(BranchEnum::FreestandingArgumentList(
	*list,
)));
```

and so on.

Which begs the question whether we actually need a From impl?
Is there a better way to implement `as_node()` and `as_mut_node()`?
In a way that prevents us from accidentally using NodeEnum::from like this..
I don't have a concrete suggestion so it's fine to leave it.

Reply:
This ties into my third future-work bullet point: `Ast` eventually shouldn't even contain a general `NodeEnum`, it should be an `AstRoot` enum, or possibly even a generic.